File: mycd.py
import sys


def compress_slashes(path: str) -> str:
    paths = path.split('/')
    count = 0

    for p in paths:
        if p == '':
            count += 1

    if count > 0:
        i = 0
        while i < count:
            paths.remove('')
            i += 1

    return '/'.join(paths)


def add_dir_on(a: str, b: str) -> str:
    a = compress_slashes(a)
    b = compress_slashes(b)

    if a == '' and b == '':
        return '/'

    a_dirs = a.split('/')
    b_dirs = b.split('/')

    final_path = []

    for d in a_dirs:
        final_path.append(d)

    for d in b_dirs:
        if d == '..' and len(final_path) != 0:
            final_path.pop()

        elif d == '..':
            continue

        elif d == '.':
            continue

        elif not d.isalnum():
            return d + ": No such file or directory"

        else:
            final_path.append(d)

    count = 0

    for p in final_path:
        if p == '':
            count += 1

    if count > 0:
        i = 0
        while i < count:
            final_path.remove('')
            i += 1

    return '/' + '/'.join(final_path)


assert add_dir_on('/', 'abc') == '/abc', '/ abc broken'
assert add_dir_on('/abc/def', 'ghi') == '/abc/def/ghi', '/abc/def ghi broken'
assert add_dir_on('/abc/def', '..') == '/abc', '/abc/def .. broken'
assert add_dir_on('/abc/def', '../..') == '/', '/abc/def ../.. broken'
assert add_dir_on('/abc/def', '../../..') == '/', '/abc/def ../../.. broken'
assert add_dir_on('/abc/def', '.') == '/abc/def', '/abc/def . broken'
assert add_dir_on('/abc/def', '..klm') == '..klm: No such file or directory', '/abc/def ..klm broken'
assert add_dir_on('/abc/def', '......') == '......: No such file or directory', '/abc/def ...... broken'
assert add_dir_on('/abc/def', '../gh///../klm/.') == '/abc/klm', '/abc/def ../gh///../klm/. broken'


def move_to(b: str) -> str:
    return add_dir_on('/', b)


assert move_to('/abc') == '/abc', 'abc/def /abc broken'
assert move_to('/abc/klm') == '/abc/klm', 'abc/def /abc/klm broken'
assert move_to('//////') == '/', '/abc/def ////// broken'

if __name__ == '__main__':

    if len(sys.argv) != 3:
        print(sys.argv[0] + " requires two strings")
        exit(1)

    if sys.argv[2][0] != '/':
        print(add_dir_on(sys.argv[1], sys.argv[2]))

    else:
        print(move_to(sys.argv[2]))

File: Mycd.java
public class Mycd {
    public static void main(String[] args) {
        if (args.length != 2) {
            System.out.println("Please input two strings");
            System.out.println(args.length);
            return;
        }

        if (args[1].charAt(0) != '/') {
            System.out.println(addDirOn(args[0], args[1]));
        }

        else {
            System.out.println(addDirOn("/", args[1]));
        }
    }

    private static String addDirOn(String arg1, String arg2) {
        arg1 = compressSlashes(arg1);
        arg2 = compressSlashes(arg2);

        if (arg1.equals("") && arg2.equals("")) {
            return "/";
        }

        String [] arg1_dirs = arg1.split("/");
        String [] arg2_dirs = arg2.split("/");

        String [] final_path = new String[arg1_dirs.length + arg2_dirs.length];

        int path_index = 0;
        for (String d : arg1_dirs) {
            final_path[path_index] = d;
            path_index++;
        }

        for (String d : arg2_dirs) {
            if (d.equals("..") && path_index > 0) {
                path_index--;
                final_path[path_index] = null;
                continue;
            }

            if (d.equals("..") && path_index == 0) {
                final_path[path_index] = null;
                continue;
            }

            else if (d.equals("..")) {
                continue;
            }

            else if (d.equals(".")) {
                continue;
            }

            for (int i = 0; i < d.length(); i++) {
                if (!Character.isLetterOrDigit(d.charAt(i))) {
                    return d + ": No such file or directory";
                }
            }

            final_path[path_index] = d;
            path_index++;
        }

        int count = 0;

        for (String p : final_path) {
            if (p != null && p.equals("")) {
                count++;
            }
        }

        String[] done = new String[final_path.length - count];

        int j = 0;
        for (String p : final_path) {
            if (p != null && !p.equals("")) {
                done[j] = p;
                j++;
            }
        }

        int nulls = 0;
        for (String p : done) {
            if (p == null) {
                nulls++;
            }
        }

        String[] final_string = new String[done.length - nulls];

        int k = 0;
        for (String p : done) {
            if (p != null) {
                final_string[k] = p;
                k++;
            }
        }

        return "/" + String.join("/", final_string);
    }

    private static String compressSlashes(String arg) {
        String [] paths = arg.split("/");
        int count = 0;

        for (String path : paths) {
            if (path.equals("")){
                count++;
            }
        }

        String [] shrunken_paths = new String[paths.length - count];

        int j = 0;
        for (String path : paths) {
            if (!path.equals("")) {
                shrunken_paths[j] = path;
                j++;
            }
        }

        return String.join("/", shrunken_paths);
    }
}
